src/main/scala/calc/Compiler.scala

```diff
@@ -66,17 +80,278 @@ object Compiler {
+      case id@IdentT(name, _) if !ps.contains(id) =>
+        Set(id.name)
+
+      case IdentT(_, _) =>
+        Set.empty
+
+      case LiteralT(_, _) =>
+        Set.empty
+
+      case BinaryOpT(_, lhs, rhs, _) =>
+        freeVarsAux(lhs) ++ freeVarsAux(rhs)
+
+      case LetT(_, value, b, _) =>
+        freeVarsAux(value) ++ freeVarsAux(b)
+
+      case IfT(cond, thenp, elsep, _) =>
+        freeVarsAux(cond) ++ freeVarsAux(thenp) ++ freeVarsAux(elsep)
+
+      case CallT(fun, args, _) =>
+        freeVarsAux(fun) ++ args.flatMap(freeVarsAux).toSet
+
+      case ClosureT(params, b, _)  =>
+        params.flatMap(freeVarsAux).toSet ++ freeVarsAux(b)
+    }
+
+    freeVarsAux(closure.body)
+  }
+
+  private def captureParams(captures: List[(String, TreeType)], scope: Scope)
+                           (implicit pos: Position): List[irt.ParamDef] = {
+    captures map { p =>
+      irt.ParamDef(irt.Ident(p._1), p._2,
+          mutable = false, rest = false)
+    }
+  }
+
+  private def captureRefs(captures: List[(String, TreeType)], scope: Scope)
+                         (implicit pos: Position): List[irt.VarRef] = {
+    captures map { p =>
+      irt.VarRef(irt.Ident(p._1))(p._2)
+    }
+  }
+
+  /** All the capture params and their corresponding VarRefs for the closure.
+    * It returns an awkward tuple to avoid finding the free variables 2 times. */
+  private def captures(closure: ClosureT, scope: Scope)
+                      (implicit pos: Position): (List[irt.ParamDef], List[irt.VarRef]) = {
+    val freeVs     = freeVars(closure)
+    val isFree     = (p: (String, _)) => freeVs.contains(p._1)
+    val cs         = scope.toList.filter(isFree)
+    val cParamDefs = captureParams(cs, scope)
+    val cVarRefs   = captureRefs(cs, scope)
+
+    (cParamDefs, cVarRefs)
+  }
+
+  /** Mangles the name according to the type. */
+  private def mangleName(name: String, tp: TreeType): String =
+    name ++ "__" ++ tp.typeSuff
+
+  /** Transforms the param list into param list for closure.
+    * Also, mangles the parameter names. */
+  private def paramsFromList(params: List[TypedTree])
+                            (implicit pos: Position): List[irt.ParamDef] = {
+    params map {
+      case IdentT(name, tp) =>
+        irt.ParamDef(
+          irt.Ident(mangleName(name, tp)),
+          irtpe.AnyType, // all params have to be AnyType
+          mutable = false, rest = false)
+
       case _ =>
-        throw new Exception(
-            s"Cannot yet compile a tree of class ${tree.getClass}")
+        throw new Exception(s"Non-identifier in argument position in function declaration.")
+    }
+  }
+
+  /** Produces the list of local variable declarations */
+  private def localParamVars(params: List[IdentT])
+                            (implicit pos: Position): irt.Tree = {
+    def localVar(param: IdentT): irt.VarDef = param match {
+      case IdentT(name, tp) =>
+        val typeCode = tp.typeSuff.charAt(0)
```

Comment:
This is not as generic as you would think. It will only work for doubles. So you might as well use `'D'` directly, here.